### PR TITLE
New block: Link Preview

### PR DESCRIPTION
**Pull request by ellatrix**
## What?

Adds a new block for previewing links. This could be done as a fallback for embeds, but I think this deserves to be its own block and I don't want to add complexity to the embed block and endpoints (my first attempt at this involved adding it to the embed API and block). We could add a button to convert to a link preview block if embedding fails.

This still needs some design work, and for some reason the stylesheet is not loading on the front end.

Screenshot 2023-02-05 at 23 25 16

## Why?

I've been missing this feature. I'd love it if we could automatically create this block when pasting a link on an empty line, but we need to see how this works with the embed block.

## How?

Uses the existing `url-details` endpoint. Is might be that we need to improve this, I don't know how good it is. Currently it's only used to preview links in the link editing UI.

## Testing Instructions

Create a link preview block and submit any URL.

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice PR, from the screenshots these look close. I'd add the 2px border radius and maybe a little padding.

I'm having some trouble reproducing, though. When I paste a URL on an empty paragraph and press Enter, I would expect the link preview to be used, but it's using a regular Embed block:

Screenshot 2023-02-06 at 09 39 47

This looks a new block, "Link preview":
Screenshot 2023-02-06 at 09 40 03

That seems okay if it's an embed variation, categorized with the other embeds. But key here would be that it's automatically transformed into, when OG tags are detected. Can we do that? 

And if so, what's a good URL to test? I'm not sure why but I'm having trouble finding good examples that work:

Screenshot 2023-02-06 at 09 40 41

**Comment by ellatrix** [comment]
> I'm having some trouble reproducing, though. When I paste a URL on an empty paragraph and press Enter, I would expect the link preview to be used, but it's using a regular Embed block:

Yes, I mentioned in the PR description that I'd like it to be possible to paste a URL and get a preview, but we need to figure out how it interacts with the embed block because right now we try embedding first.

> But key here would be that it's automatically transformed into, when OG tags are detected. Can we do that?

What do you mean? I think we are checking OG tags now though the `url-details` endpoint. I need to double check that.

> And if so, what's a good URL to test?

It's good that you find examples that don't work. Could you share them? Cause ideally any link should work. If there's no OG tags, then we can fall back to the document title.

**Comment by jasmussen** [comment]
> Yes, I mentioned in the PR description that I'd like it to be possible to paste a URL and get a preview, but we need to figure out how it interacts with the embed block because right now we try embedding first.

🚀 

> What do you mean? I think we are checking OG tags now though the url-details endpoint. I need to double check that.

Same as above: that it tries to convert it into your new block if there's content to show in it.

> It's good that you find examples that don't work. Could you share them? Cause ideally any link should work. If there's no OG tags, then we can fall back to the document title.

I actually tried pasting in a couple of Gutenberg PRs, such as this one. I also tried a couple of wikipedia links, such as `https://en.wikipedia.org/wiki/Atomic_Swing`:

Screenshot 2023-02-07 at 09 01 40

**Comment by ellatrix** [comment]
> I actually tried pasting in a couple of Gutenberg PRs, such as this one. I also tried a couple of wikipedia links, such as `https://en.wikipedia.org/wiki/Atomic_Swing`:

But are you trying it with the Link Preview block? The image you shared is the Embed block?

**Comment by jasmussen** [comment]
My apologies, I should've included that. Nothing happens, no console error, I click embed but it's stuck:

link preview

**Comment by getdave** [comment]
Nice PR. I particularly like the idea of auto transforming pasting links into this block 👍 

I tested the `url-details` endpoint with `https://en.wikipedia.org/wiki/Atomic_Swing` and it works reasonably well returning the following:

```
{
    "title": "Atomic Swing - Wikipedia",
    "icon": "https://en.wikipedia.org/static/favicon/wikipedia.ico",
    "description": "",
    "image": "https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Atomic_Swing_Wordmark.svg/1200px-Atomic_Swing_Wordmark.svg.png"
}
```

The REST API endpoint itself has several methods for parsing the various url details. You can see these starting at

https://github.com/WordPress/wordpress-develop/blob/7e0fe194aa7b65625af460957f9a468a5194a926/src/wp-includes/rest-api/endpoints/class-wp-rest-url-details-controller.php#L286

Most of the methods look for OG tags as well as standard HTML tags. There are certain sites which is doesn't play well with, including sites that rely on JavaScript for rendering and which also don't provide useful metadata in the server response. I don't see any way we can work around that. Google et al. presumably do it by relying on their internal search APIs to retrieve data parser by their own crawlers. We can't do that unfortunately...

**Comment by ellatrix** [comment]
This is so strange. I had it working properly when I created the PR, and now the `onSubmit` handler doesn't seem to be called any more.

**Comment by ellatrix** [comment]
@joen My apologies as well, I accidentally added an onClick listener that prevents default behaviour on the whole block, so that's why the button in the placeholder didn't work anymore. 🙈

**Review by getdave — COMMENTED** [review]
I'd like a way to edit my link after I've submitted it. Currently you can't do that.

Also I tried pasting in a YT watch link like `https://www.youtube.com/watch?v=0mUrwyyw-8I&t=269s`. The preview was broken (because many elements of data are expected in the block but missing in the response) but the strangest thing was when you click on the link it tries to load internally within the iframe.


https://user-images.githubusercontent.com/444434/217622041-a91d853a-d050-49b5-a561-b726000a747d.mp4

**Comment by ellatrix** [comment]
@getdave I added URL editing and fixed it for cases where some of the fields is empty.

**Comment by ellatrix** [comment]
Does anyone know how to get the stylesheet enqueued on the front-end?

File: packages/block-library/src/link-preview/edit.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+			setIsFetching( true );
+			__experimentalFetchUrlData( url )
+				.then( ( data ) => {
+					setAttributes( data );
+				} )
+				.finally( () => {
+					setIsFetching( false );
+				} );
+		}
+	}, [] );
```
Comment (getdave) [inline packages/block-library/src/link-preview/edit.js]: It seems this will only ever run once because there are no dependencies listed. If the `url` changes don't we want to re-fetch the `data` with using the new URL? 
```suggestion
	}, [ url, title, icon, image ] );
```

File: packages/block-library/src/link-preview/edit.js [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+								description: '',
+							} );
+							setIsFetching( true );
+							__experimentalFetchUrlData( urlValue )
+								.then( ( data ) => {
+									setAttributes( data );
+									setIsEditingUrl( false );
+								} )
+								.finally( () => {
+									setIsFetching( false );
```
Comment (getdave) [inline packages/block-library/src/link-preview/edit.js]: Do we need to stop editing here also?

```suggestion
									setIsFetching( false );
									setIsEditingUrl( false );
```

File: packages/block-library/src/link-preview/edit.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+						onSubmit={ () => {
+							setAttributes( {
+								url: urlValue,
+								title: '',
+								icon: '',
+								image: '',
+								description: '',
+							} );
+							setIsFetching( true );
+							__experimentalFetchUrlData( urlValue )
```
Comment (getdave) [inline packages/block-library/src/link-preview/edit.js]: We may not need a fetch here if we put the effectful code in the `useEffect` and have a single place to fetch. It also makes the block attributes the single source of truth for the "link". In this case we're using the temporary local state variable as the url.

The fact you went down this route makes me think I've missed something and it needed to be set up like this though... 🤔

File: packages/block-library/src/link-preview/edit.js [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+						<input
+							type="url"
+							value={ urlValue }
+							className="components-placeholder__input"
+							aria-label={ __( 'URL' ) }
+							placeholder={ __( 'Enter URL to embed here…' ) }
+							onChange={ ( event ) => {
+								setURLValue( event.target.value );
+							} }
+						/>
```
Comment (getdave) [inline packages/block-library/src/link-preview/edit.js:102]:
```suggestion
<InputControl
	label={ __( 'URL' ) }
	hideLabelFromVision
	className="components-placeholder__input"
	type="url"
	value={ urlValue }
	placeholder={ __( 'Enter URL to embed here…' ) }
	onChange={ ( event ) => {
		setURLValue( event.target.value );
	} }
/>
```

We can use the component here if we want to inherit the standard components package styling.

File: packages/block-library/src/link-preview/edit.js [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+							setIsEditingUrl( true );
+							setURLValue( url );
+						} }
+					/>
+				</ToolbarGroup>
+			</BlockControls>
+
+			{ isFetching && <Spinner /> }
+
+			{ ! isFetching && (
```
Comment (getdave) [inline packages/block-library/src/link-preview/edit.js]: I pushed this change so we see a spinner whilst fetching rather than half of the preview.

We should also consider what to show if the rich url data is not available for the given link. Some kind of fallback for each attribute perhaps? Currently it could look like an error. 

Instead we should try to communicate to the user that we are missing some of the data.

File: packages/block-library/src/link-preview/save.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+				{ title && <strong>{ title }</strong> }
+				{ icon && (
+					<img
+						className="link-preview__icon"
+						src={ icon }
+						alt={ new URL( url ).host }
+					/>
+				) }
+				{ title ? new URL( url ).host.replace( /^www\./, '' ) : url }
+			</div>
```
Comment (getdave) [inline packages/block-library/src/link-preview/save.js]: Same comment here as for `edit` - let's provide fallbacks if data is missing.

Comment (getdave) [reply packages/block-library/src/link-preview/save.js]: Also the code here could be shared with the `edit` file.

File: packages/block-library/src/link-preview/transforms.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
+/**
+ * WordPress dependencies
+ */
+import { createBlock } from '@wordpress/blocks';
+
+/**
+ * Default transforms for generic embeds.
+ */
+const transforms = {
```
Comment (getdave) [inline packages/block-library/src/link-preview/transforms.js:9]: I tried pasting `https://aheadcreative.com` into the editor canvas and it turned it into an embed block. I couldn't see how to get the link preview block to work.

**Review by getdave — COMMENTED** [review]
Pushed one change ( hope you don't mind?) and left some comments to consider.

There are some visual aspects that it would be nice to tidy up too including providing fallbacks in case of missing data. You can see how we handled that for `LinkControl` in 

https://github.com/WordPress/gutenberg/blob/7b3c06edd3ee3e8573819176dd8f426774333d04/packages/block-editor/src/components/link-control/link-preview.js

File: packages/block-library/src/link-preview/transforms.js (hunk 7 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/transforms.js:9]: Oh, I've not added that yet. When should we convert a pasted link to a link preview? When it fails to embed?

File: packages/block-library/src/link-preview/save.js (hunk 6 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/save.js]: I already added a fallback: the fallback is simply the URL.

File: packages/block-library/src/link-preview/edit.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js]: I think it's ok if either the icon or image is not available. Maybe we could make the title a requirement because every web page should have at least a title.

**Comment by ellatrix** [comment]
Regarding paste, I'm wondering if we should give people the option between three different blocks when pasting a link:
* An embed
* A link preview block
* Just a paragraph with the link

So upon paste, we could show a placeholder with the three different blocks and a preview of them from which the user can pick. Not sure if this is a good or bad idea.

**Comment by jasmussen** [comment]
> Regarding paste, I'm wondering if we should give people the option between three different blocks when pasting a link:

Just to be sure: is the Link Preview superior to the current Link Embed block? If it is, it could replace it (or the embed could be "refactored" into this). Or is there a reason to keep the existing link embed block around?

That said, one way of transforming one into the other could potentially be this top of the inspector interface that we are discussing for navigation items in #46195. I.e. the toggles. Maybe?

File: packages/block-library/src/link-preview/edit.js (hunk 5 above)
Comment (getdave) [reply packages/block-library/src/link-preview/edit.js]: ...it _should_ but the endpoint is obviously struggling with certain sites. I guess all I'm suggesting is we provide a visual fallback for `icon` and `image` so the layout is retained. It's just a suggestion though. I should probably defer to @jasmussen on that one 🙇

File: packages/block-library/src/link-preview/transforms.js (hunk 7 above)
Comment (getdave) [reply packages/block-library/src/link-preview/transforms.js:9]: If that can be achieved then yes that would seem logical 👍

**Comment by ellatrix** [comment]
@jasmussen What is the Link Embed? I've never heard of it and I'm also not seeing it in the block library.

File: packages/block-library/src/link-preview/edit.js (hunk 4 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js:102]: It does have normal styling right now. Why are other blocks using `input` over `InputControl`? I just copied this from the embed block.

File: packages/block-library/src/link-preview/edit.js (hunk 5 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js]: Made the title a requirement.

File: packages/block-library/src/link-preview/edit.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js]: I removed the effect. We shouldn't be doing any fetching that doesn't come from a user interaction since we're saving the content as block attributes. If there's no url or title, we automatically switch to edit mode.

File: packages/block-library/src/link-preview/edit.js (hunk 2 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js]: We shouldn't stop editing if there's a problem. I added error handling, but it still needs some polish design wise.

File: packages/block-library/src/link-preview/edit.js (hunk 1 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js]: Removed the effect, see below.

File: packages/block-library/src/link-preview/edit.js (hunk 3 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js]: Restored the effect so it can be transformed to from an embed block

**Comment by ellatrix** [comment]
Not sure how to handle paste. When an embed block is created, there's no way to know it came from paste, and we don't just want to replace any failing embed block to a link preview (for example when you load a post in the editor). We only want it on paste. The paste handler also doesn't know when a block "failed".

It seems that the only option right now is to offer it next to "Convert to link" in the placeholder you see when an Embed failed to preview.

In any case, I don't think this should be a blocker for adding this new block?

Would love some help with the design of this preview.

File: packages/block-library/src/link-preview/edit.js [hunk 8]
```diff
@@ -35,7 +36,34 @@ export default function LinkPreviewEdit( props ) {
...
 	} );
 
-	if ( isEditingUrl ) {
+	useEffect( () => {
+		if ( url && ! title ) {
+			setIsFetching( true );
+			__experimentalFetchUrlData( url )
+				.catch( () => {
+					setHasError( true );
+				} )
```
Comment (getdave) [inline packages/block-library/src/link-preview/edit.js:45]: Out of curiosity is putting `.catch` before `.then`  purely syntatical? My understanding is that the Promise will resolve/reject with `then/catch` being called accordingly no matter the order, but interested to know if my mental model is flawed.

File: packages/block-library/src/link-preview/edit.js (hunk 4 above)
Comment (getdave) [reply packages/block-library/src/link-preview/edit.js:102]: Mainly just noticing that normalising towards our component library will ensure we have standard styling across all blocks. 

Not a blocker.

**Review by getdave — COMMENTED** [review]
The single `fetch` in the effect is much better 👍 

Tested with a super slow network and noticed that the form remains visible even during fetching. This means that it's possible to update the URL after the initial fetch has been initiated. This leads to potential network race conditions.

We should probably disable the UI whilst fetching - what do you think?

https://user-images.githubusercontent.com/444434/220888653-d5472b78-8404-4615-aa04-38b9cd32fd64.mp4

I also pushed a change to improve the error notice styling:

Screen Shot 2023-02-23 at 11 14 51

**Comment by getdave** [comment]
> In any case, I don't think this should be a blocker for adding this new block?

Not a blocker.

File: packages/block-library/src/link-preview/edit.js (hunk 8 above)
Comment (ellatrix) [reply packages/block-library/src/link-preview/edit.js:45]: Order doesn't matter :)

Comment (getdave) [reply packages/block-library/src/link-preview/edit.js:45]: Glad I didn't have to completely relearn my mental model 👍

**Comment by jasmussen** [comment]
> @jasmussen What is the Link Embed? I've never heard of it and I'm also not seeing it in the block library.

Sorry, just coming back to this now. What I mean is, currently when you paste a link on a new line and press enter, it's transformed into a sort of Oembed block that does show a little preview. That's what I called the "Link Embed", but I'm not actually sure what technically it's called — perhaps just the regular Embed block? The main point I wanted to make was that this feature already appears to exist, even if it's hit and miss, and that the superior version should replace whatever is in place at the moment.

Does that make sense?

**Comment by ellatrix** [comment]
@jasmussen That sort of what we talked about before. Whether this should serve as a fallback if embedding doesn't work. It's two different things though. The current embed block is used for video embedding etc., but not every link can be embedded. This is different: it fetches the title and featured image of a page so you can "preview" the link. I think it's worth starting by adding this block and then figuring out when a pasted link should be embedded or previewed.

**Comment by jasmussen** [comment]
I'm mainly wondering about the flow. If I paste a link in Facebook, Twitter, Slack, or otherwise, and give it its own line, I expect a visual preview/embed. If we build a new block that provides a better preview than what _occasionally_ works here, that's cool, but if I have to manually pick that particular block first in order to accomplish that, I worry no-one will discover it, due to the expected embedding flow.

**Comment by Mamaduka** [comment]
@ellatrix, I've got a block plugin that does the same. Feel free to use any part of it - https://github.com/Mamaduka/bookmark-card.

> The single fetch in the effect is much better 👍

I don't think that the fetch should live in effect. The block only wants to retrieve data on a particular interaction: when users press the button. I think it would be more suitable to move that logic inside submit event handler.

**Comment by masteradhoc** [comment]
@ellatrix any update on this block? looks promising. thanks for the great work!

**Comment by getdave** [comment]
> > The single fetch in the effect is much better 👍
> 
> I don't think that the fetch should live in effect. The block only wants to retrieve data on a particular interaction: when users press the button. I think it would be more suitable to move that logic inside submit event handler.

@Mamaduka Just for [context on why this was implemented](https://github.com/WordPress/gutenberg/pull/47765#discussion_r1114246756). Is there a particular reason why you dislike the `useEffect`? I assume you have some learnings from your existing Plugin to share? Thanks!

**Comment by Mamaduka** [comment]
Thanks for sharing the context, @getdave!

I don't dislike the `useEffect` hook, but I think there are multiple places in the project where we're misusing them. While I understand that it's hard to follow recommended practices in large projects like Gutenberg, we should be cautious when introducing new cases.

I think @kevin940726 summarizes the "cons" well in this comment - https://github.com/WordPress/gutenberg/issues/48225#issuecomment-1510631497.

> Restored the effect so it can be transformed to from an embed block

We can also display the placeholder component with a URL and let the users "embed" the links, instead of auto-embedding them.

**Comment by kevin940726** [comment]
> I don't dislike the `useEffect` hook, but I think there are multiple places in the project where we're misusing them. While I understand that it's hard to follow recommended practices in large projects like Gutenberg, we should be cautious when introducing new cases.

I fully agree with @Mamaduka here. The `useEffect` here looks a bit like an anti-pattern and could cause some unexpected and difficult-to-fix side effects if we aren't careful.

If it's the only way that could be done to support transforming from and to an embed block, then at least we should document it with a comment so that people don't blindly copy-paste it.

File: packages/block-library/src/link-preview/edit.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+				.then( ( data ) => {
+					if ( ! data || ! data.title ) {
+						setHasError( true );
+					} else {
+						setHasError( false );
+						setIsEditingUrl( false );
+						setAttributes( {
```
Comment (kevin940726) [inline packages/block-library/src/link-preview/edit.js:52]: I think we also need a couple of things if this effect is required.

- Add `__unstableMarkNextChangeAsNotPersistent` before `setAttributes` to prevent "[undo traps](https://github.com/WordPress/gutenberg/issues/8119#issuecomment-1185238067)".
- Add a cleanup function to prevent race conditions and memory leaks. Don't update states if the component is already unmounted.

**Comment by getdave** [comment]
Just remembered this PR. Is there still interest in pursuing this block?

I think we could get it to land by descoping this PR a little even if that compromises initial utility. 

For example by just shipping the block code and not including the transforms and auto-embedding behaviour.

That could all be handled in follow ups.

Looks like we need to address a few nits as well.

**Comment by ellatrix** [comment]
@getdave I thought it's already pretty descoped. What were you thinking of removing? I would still love this block. :)

**Comment by getdave** [comment]
> ...not including the transforms and auto-embedding behaviour.

@ellatrix I was thinking this was more complex than it actually is.

So nothing to descope. Just fixing up the effectful code. Probably one for post 6.5.
